hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/volume/OMVolumeSetOwnerRequest.java

```diff
@@ -171,7 +171,8 @@ public OMClientResponse validateAndUpdateCache(OzoneManager ozoneManager,
       omResponse.setSetVolumePropertyResponse(
           SetVolumePropertyResponse.newBuilder().build());
       omClientResponse = new OMVolumeSetOwnerResponse(oldOwner,
-          oldOwnerVolumeList, newOwnerVolumeList, omVolumeArgs,
+          oldOwnerVolumeList, newOwnerVolumeList,
+          (OmVolumeArgs) omVolumeArgs.clone(),
```

Comment:
It is not clear to me why an extra clone of omVolumeArgs is needed here. omVolumeArgs is just created a few line above in Line 129. Since it is not an argument provided by the caller and not handed over to other clients, can you post stack tracing showing concurrent modification exception?

Reply:
https://issues.apache.org/jira/browse/HDDS-2322
The issue is seen with Key Operations, I think we might see the same issue with Other Ops too. Let's take a case where RemoveAcl submitted OMVolumeArgs to flushThread and other thread like removeAcl is updating same OmVolumeArgs on this. Then when we might see ConcurrentModificationException. So to cover these kinds of scenario's updated all the places where we submit the response to doubleBuffer flush threads.

Reply:
@xiaoyuyao does this address your concern?

Reply:
bq. other thread which is performing removeAcl is updating the same OmVolumeArgs on this(Like we have a list and map internally for acls)

We've already done defensive copy for get Java object of Volume/Bucket/Key from DB and converted the client provided acl from protobuf when creating OMxxxAclRequest. I think we should trace down the real cause of the CME here. Generally, I agree making object immutable is a good thing. However, unnecessary clone object has performance implications  that we should be cautious. 

bq. when flush happens we might flush these entries to rocksdb when some other operation is changing the same entry which might be still submitted/not submitted to doublebuffer flush thread.

omVolumeArgs/KeyInfo/BucketInfo are created per request from DB like below:
     omVolumeArgs = omMetadataManager.getVolumeTable().get(dbVolumeKey);

Can you elaborate on the operation that could change the per request created volume/bucket/key object?

Reply:
Yes but with cache enabled, every time it is not a new object.

Reply:
Thank You @xiaoyuyao for the offline discussion. Posted a patch to make immutable value from get() API when it is returned from cache. (In this updated only for Bucket and Volume requests) Updated bucket request also in this, as that got checked in as part of HDDS-2333. Please share your view on this.